streams/src/main/java/org/apache/kafka/streams/state/internals/ThreadCache.java

```diff
@@ -122,13 +122,17 @@ public void put(final String namespace, byte[] key, LRUCacheEntry value) {
     }
 
     public LRUCacheEntry putIfAbsent(final String namespace, byte[] key, LRUCacheEntry value) {
-        final NamedCache cache = getOrCreateCache(namespace);
-        return cache.putIfAbsent(Bytes.wrap(key), value);
+        final LRUCacheEntry originalValue = get(namespace, key);
```

Comment:
This method is not `synchronized`. So there could be a race condition here. I think it should be:
```
final NamedCache cache = getOrCreateCache(namespace);
final LRUCacheEntry result = cache.putIfAbsent(Bytes.wrap(key), value);
maybeEvict(namespace);
return result;
```